[mlir][Pass] Add new FileTreeIRPrinterConfig

christopherbate opened the pull request:
This change expands the existing instrumentation that prints the IR
before/after each pass to an output stream (usually stderr). It adds
a new configuration that will print the output of each pass to a
separate file. The files will be organized into a directory tree
rooted at a specified directory. For existing tools, a CL option
`-mlir-print-ir-tree-dir` is added to specify this directory and
activate the new printing config.

The created directory tree mirrors the nesting structure of the IR. For example,
if the IR is congruent to the pass-pipeline
"builtin.module(pass1,pass2,func.func(pass3,pass4),pass5)", and
`-mlir-print-ir-tree-dir=/tmp/pipeline_output`, then then the tree file tree
created will look like:

```
/tmp/pass_output
├── builtin_module_the_symbol_name
│   ├── 0_pass1.mlir
│   ├── 1_pass2.mlir
│   ├── 2_pass5.mlir
│   ├── func_func_my_func_name
│   │   ├── 1_0_pass3.mlir
│   │   ├── 1_1_pass4.mlir
│   ├── func_func_my_other_func_name
│   │   ├── 1_0_pass3.mlir
│   │   ├── 1_1_pass4.mlir
```

The subdirectories are named by concatenating the relevant parent
operation names and symbol name (if present). The printer keeps a
counter associated with ops that are targeted by passes and their
isolated-from-above parents. Each filename is given a numeric prefix
using the counter value for the op that the pass is targeting and then
prepending the counter values for each parent. This gives a naming
where it is easy to distinguish which passes may have run concurrently
vs. which have a clear ordering. In the above example, for both
`1_1_pass4.mlir` files, the first `1` refers to the counter for the
parent op, and the second refers to the counter for the respective
function.

joker-eph commented:
This is really cool, shall we add some doc? The `print-after-all` is documented here: https://mlir.llvm.org/docs/PassManagement/#ir-printing

jpienaar commented:
+1 this looks really nice, and easy to follow/reproduce from disk.

christopherbate commented:
Thanks for the comments! I updated the docs as well.

joker-eph commented on mlir/include/mlir/Pass/PassManager.h:
```diff
@@ -381,6 +381,42 @@ class PassManager : public OpPassManager {
+  /// ├── builtin_module_the_symbol_name
+  /// │   ├── 0_pass1.mlir
+  /// │   ├── 1_pass2.mlir
+  /// │   ├── func_func_my_func_name
+  /// │   │   ├── 2_pass3.mlir
+  /// │   │   ├── 3_pass4.mlir
+  /// │   ├── func_func_my_other_func_name
+  /// │   │   ├── 4_pass3.mlir
+  /// │   │   ├── 5_pass4.mlir
```
```suggestion
  /// │   ├── func_func_my_func_name
  /// │   │   ├── 2_passA.mlir
  /// │   │   ├── 5_passB.mlir
  /// │   ├── func_func_my_other_func_name
  /// │   │   ├── 3_passA.mlir
  /// │   │   ├── 4_passB.mlir
```

Use names instead of numbers for the passes.

Also I'd use counter that aren't in sequence to illustrate that parallelism makes this non-deterministic.
And by the way, that seems kind of an issue, I wonder if we could do better, for example prefix instead:

```
├── builtin_module_the_symbol_name
   /// │   ├── 0_pass1.mlir
   /// │   ├── 1_pass2.mlir
   /// │   ├── func_func_my_func_name
   /// │   │   ├── 0_1_pass3.mlir
   /// │   │   ├── 0_2_pass4.mlir
   /// │   ├── func_func_my_other_func_name
   /// │   │   ├── 1_1_pass3.mlir
   /// │   │   ├── 1_2_pass4.mlir
```

That is every time we get into a new nesting we prepend with a prefix that is the operation position in the parent (here `0` and `1` for each function) and restart with a local counter instead of a shared one.

christopherbate replied:
Wouldn't the below two pipelines both produce the same file tree (and same file names) using your suggested naming?

1. `builtin.module(cse,cse,func.func(cse))`
2. `builtin.module(cse,func.func(cse),cse)`

joker-eph replied:
Aren't the numbers prefix here to disambiguate in my example?
(not the two different counters I mentioned.

christopherbate replied:
NVM I see what you mean now. misunderstood what you meant by " the operation position in the parent".

joker-eph reviewed (commented):
Is the nesting working with dynamic pass pipelines?

jpienaar commented:
Is this just pending the dynamic pass pipeline discussion?

christopherbate commented:
Sorry this fell off my TODO list, I plan on picking it back up next week.

christopherbate committed:
[Pass] Add new FileTreeIRPrinterConfig

This change expands the existing instrumentation that prints the IR
before/after each pass to an output stream (usually stderr). It adds
a new configuration that will print the output of each pass to a
separate file. The files will be organized into a directory tree
rooted at a specified directory. For existing tools, a CL option
`-mlir-print-ir-tree-dir` is added to specify this directory and
activate the new printing config.

The created directory tree mirrors the nesting structure of the IR. For example,
if the IR is congruent to the pass-pipeline
"builtin.module(pass1,pass2,func.func(pass3,pass4),pass5)", and
`-mlir-print-ir-tree-dir=/tmp/pipeline_output`, then then the tree file tree
created will look like:

```
/tmp/pass_output
├── builtin_module_the_symbol_name
│   ├── 0_pass1.mlir
│   ├── 1_pass2.mlir
│   ├── 2_pass5.mlir
│   ├── func_func_my_func_name
│   │   ├── 1_0_pass3.mlir
│   │   ├── 1_1_pass4.mlir
│   ├── func_func_my_other_func_name
│   │   ├── 1_0_pass3.mlir
│   │   ├── 1_1_pass4.mlir
```

The subdirectories are named by concatenating the relevant parent
operation names and symbol name (if present). The printer keeps a
counter associated with ops that are targeted by passes and their
isolated-from-above parents. Each filename is given a numeric prefix
using the counter value for the op that the pass is targeting and then
prepending the counter values for each parent. This gives a naming
where it is easy to distinguish which passes may have run concurrently
vs. which have a clear ordering. In the above example, for both
`1_1_pass4.mlir` files, the first `1` refers to the counter for the
parent op, and the second refers to the counter for the respective
function.

christopherbate committed:
Fix docs with updated naming scheme

joker-eph commented on mlir/docs/PassManagement.md:
```diff
@@ -1370,22 +1370,32 @@ func.func @simple_constant() -> (i32, i32) {
 ├── builtin_module_the_symbol_name
 │   ├── 0_pass1.mlir
 │   ├── 1_pass2.mlir
+│   ├── 2_pass5.mlir
```
Somehow I would want the counter here to be 3, and 2 being used for the nested pass manager.
But we can leave this up for a future revision!

joker-eph reviewed (approved):
Thanks @christopherbate ! :)

christopherbate committed:
fix doc comment to reflect prefix naming scheme

christopherbate commented:
Windows build fails due to unrelated issue 

https://buildkite.com/llvm-project/github-pull-requests/builds/66880#018fa910-2263-4ed2-91fe-cff2dbf95616/394-6441


FAILED: tools/flang/lib/Parser/CMakeFiles/obj.FortranParser.dir/expr-parsers.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Parser -IC:\ws\src\flang\lib\Parser -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\expr-parsers.cpp.obj /Fdtools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\ /FS -c C:\ws\src\flang\lib\Parser\expr-parsers.cpp
  | C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">

FAILED: tools/flang/lib/Parser/CMakeFiles/obj.FortranParser.dir/expr-parsers.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Parser -IC:\ws\src\flang\lib\Parser -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\expr-parsers.cpp.obj /Fdtools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\ /FS -c C:\ws\src\flang\lib\Parser\expr-parsers.cpp
  | C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">


FAILED: tools/flang/lib/Parser/CMakeFiles/obj.FortranParser.dir/expr-parsers.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Parser -IC:\ws\src\flang\lib\Parser -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\expr-parsers.cpp.obj /Fdtools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\ /FS -c C:\ws\src\flang\lib\Parser\expr-parsers.cpp
  | C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">


FAILED: tools/flang/lib/Parser/CMakeFiles/obj.FortranParser.dir/expr-parsers.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Parser -IC:\ws\src\flang\lib\Parser -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\expr-parsers.cpp.obj /Fdtools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\ /FS -c C:\ws\src\flang\lib\Parser\expr-parsers.cpp
  | C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">


FAILED: tools/flang/lib/Parser/CMakeFiles/obj.FortranParser.dir/expr-parsers.cpp.obj
  | sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\flang\lib\Parser -IC:\ws\src\flang\lib\Parser -IC:\ws\src\flang\include -Itools\flang\include -Iinclude -IC:\ws\src\llvm\include -IC:\ws\src\llvm\..\mlir\include -Itools\mlir\include -Itools\clang\include -IC:\ws\src\llvm\..\clang\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Fotools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\expr-parsers.cpp.obj /Fdtools\flang\lib\Parser\CMakeFiles\obj.FortranParser.dir\ /FS -c C:\ws\src\flang\lib\Parser\expr-parsers.cpp
  | C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">



C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">

C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xatomic.h(17): fatal error C1060: compiler is out of heap space
 

<br class="Apple-interchange-newline">
